fs/btrfs/zstd.c

```diff
@@ -95,7 +114,26 @@ static int zstd_compress_pages(struct list_head *ws,
 	unsigned long len = *total_out;
 	const unsigned long nr_dest_pages = *out_pages;
 	unsigned long max_out = nr_dest_pages * PAGE_SIZE;
-	ZSTD_parameters params = zstd_get_btrfs_parameters(len);
+
+	ZSTD_parameters params = zstd_get_btrfs_parameters(len, workspace->level);
+	size_t new_size = ZSTD_CStreamWorkspaceBound(params.cParams);
+	if (new_size > workspace->size) {
+		if (new_size < workspace->lowest_failed_size) {
+			void *new_mem = kvmalloc(new_size, GFP_KERNEL);
+			if (new_mem) {
+				kvfree(workspace->mem);
+				workspace->mem = new_mem;
+				workspace->size = new_size;
+			} else {
+				ZSTD_parameters original_params = zstd_get_btrfs_parameters(len, ZSTD_BTRFS_DEFAULT_LEVEL);
```

Comment:
Should we also save the current params in the workspace? That way if we fail we can base it off of the current params, instead of going back to level 3.

```c
zstd_set_level(ws, 10); // success
zstd_set_level(ws, 15); // failure
```